### ISSUE TITLE
does buddy_resize invalidates all previously allocated memory pointers?

### ISSUE THREAD

**Issue by janedoe-lab** [spaskalev/buddy_alloc#128]
I am not sure if this is bug I am experiencing or this is working as designed.

I am using this library to manage a dynamically sized memory arena. Whenever buddy_alloc fails, I increase amount of commited memory to the arena and then call buddy_resize(). In small tests it seems like it works as I expect, but in larger application I see weird memory rewrites over already allocated data and can't quite figure out if this is due to buddy_resize() changing the layout, or there is some subtle bug/race-condition in my larger code-base.

Sorry, can't give an test case, since all my smaller tests passing just fine.

**Comment by spaskalev** [comment]
Hi and thank you for the interest in the project! Resizing the arena is probably the most complex operation in the entire allocator and while it is tested I would not be surprised if there is some subtle bug with it. In fact I had to create the check invariant function to detect and fix the initial issues with the resize.

The resize itself should not alter the allocation layout in any way - its just adding/removing capacity on the right side of the allocator and accounting for virtual slots (those that don't fall in the given memory) and the metadata itself if your are using the allocator in the embedded mode.

Can you please share the mode of operation and arena sizes that you are using - like external metadata or embedded metadata and chaging from what arena size to what arena size. Also, what's the used alignment (if not the default) ?

**Comment by spaskalev** [comment]
Can you also share your routine that adds memory to the arena ? If you are using `realloc` from the standard librarythat can give you a different base pointer and then all of your previous allocations will be invalid.

**Comment by janedoe-lab** [comment]
Thank you for confirming my expectation that already allocated data should not, at least by design, change. For now I just assume there is a subtle bug somewhere in my code.

I am using combination of VirtualAlloc to reserve a large piece of memory (512MB), and later calling VirtualAlloc to commit more memory from that reserved area (starting with 16Mb and then doubling it every resize), so the base memory address does not change.

The project I am working on is one big mess of hooks into Windows' API and I suspect there is race condition somewhere. I will keep debugging and if I will be able to make a small replication case I will certainly post it.

Thank you so much!!

**Comment by spaskalev** [comment]
If you get a reproducer please let me know. I'll devise some further tests on the resizing functionality, perhaps there's still something lurking in there.

**Comment by janedoe-lab** [comment]
I think I found the source of the issue I was sporadically experiencing. As part of my allocator built on top of buddy_alloc, I try to allocate requested memory, and if it fails - try to upsize arena size. But if I reach max arena size limit, I call again buddy_resize(buddy, old_arena_size) to free up memory after unsuccessful allocation. I've replicated it using this code:

```
#include <stdlib.h>

#define BUDDY_ALLOC_IMPLEMENTATION
#include "buddy_alloc.h"
#undef BUDDY_ALLOC_IMPLEMENTATION

int main()
{
    void *str1, *str2;

    size_t arena_size = 65536;

    void *buddy_metadata = malloc(buddy_sizeof(arena_size));
    void *buddy_arena = malloc(arena_size);

    struct buddy *buddy = buddy_init(buddy_metadata, buddy_arena, 1024);

    str1 = buddy_malloc(buddy, 128);
    fprintf(stderr, "str1: %p\n", str1);

    buddy_resize(buddy, 2048);
    buddy_resize(buddy, 1024); // this triggers error

    str2 = buddy_malloc(buddy, 256);
    fprintf(stderr, "str2: %p\n", str2);

    free(buddy_metadata);
    free(buddy_arena);
}
```

Example output:

```
str1: 0x564982da5550
str2: 0x564982da5550
```

Notice how both pointers point to the same memory address.

If I remove `buddy_resize(buddy, 1024)` line, I get expected different addresses in the output:

```
str1: 0x55829e4da550
str2: 0x55829e4da650
```

Not sure if this is bug or expected behavior. For now I will just disable downsizing arena after failure of allocating memory and will try to use some heuristic to predict when `buddy_malloc` might fail with available arena size. But ideally I want to be able to downsize the arena if I know for sure that no memory was allocated between calls to `buddy_resize()`

**Comment by spaskalev** [comment]
You're right, adding calls to `buddy_debug(buddy)` before, between and after the resize operations show that the allocated slot `pos index: 8 pos depth: 4 status: 2 bitset-len: 2 bitset-at: 25 size: 128` before the resize becomes `pos index: 16 pos depth: 5 status: 2 bitset-len: 2 bitset-at: 56 size: 128` after the upsize and then it becomes `pos index: 8 pos depth: 4 status: 0 bitset-len: 2 bitset-at: 25` after the downsize.

Hence, this is clearly a bug in the allocator :) Thanks for reporting it again and coming up with a reproducer, I really appreciate it. I'll dig in and fix it.

**Comment by spaskalev** [comment]
This is caused by the `bitset_clear_range(bitset, bitset_range(length, length+by-1));` line in the `bitset_shift_left` function. Comment it out and your test passes, along with all other tests in the suite (except for test_bitset_shift ;))

I'll come up with a proper fix but meanwhile you just drop it and let me know how it goes.

**Comment by janedoe-lab** [comment]
It appears to work after I commented out that line. Thank you!

**Comment by spaskalev** [comment]
The proper fix is in now :) Please me know if you have any issues. Thanks!
